engines/private/code.cpp

```diff
@@ -0,0 +1,426 @@
+#include "common/str.h"
+#include "common/debug.h"
+#include "common/hash-ptr.h"
+
+#include "grammar.h"
+#include "grammar.tab.h"
+#include "private.h"
+
+namespace Private {
+
+Setting *setting;
+SettingMap settingMap;
+
+Datum	*stack       = NULL;	/* the stack */
+Datum	*stackp      = NULL;	/* next free spot on stack */
+
+Inst	*prog        = NULL;	/* the machine */
+Inst	*progp       = NULL;	/* next free spot for code generation */
+Inst	*pc          = NULL;	/* program counter during execution */
+
+
+static struct InstDescr {
+    const Inst func;
+    const char *name;
+} instDescr[] = {
+    { 0,        "STOP",     },
+    { constpush,"constpush" },
+    { strpush,  "strpush",	},
+    { varpush,  "varpush",	},
+    { funcpush, "funcpush",	},
+    { eval,	    "eval",     },
+    { ifcode,   "ifcode",	},
+    { add,      "add",      },
+    { negate,	"negate",	},
+
+    { 0, 0}
+};
+
+PtrToName _insts;
+
+void initInsts() {
+    for (InstDescr *fnc = instDescr; fnc->name; fnc++) {
+        _insts[(void *)fnc->func] = new Common::String(fnc->name);
+    }
+}
+
+/* initialize for code generation */
+void initSetting() {
+    setting = (Setting*) malloc(sizeof(Setting));
+    memset((void *) setting, 0, sizeof(Setting));
+
+    prog = (Inst *) &setting->prog;
+    stack = (Datum *) &setting->stack;
+
+    stackp = stack;
+    progp = prog;
+}
+
+void saveSetting(char *name) {
+    Common::String s(name);
+    settingMap.setVal(s, setting);
+}
+
+void loadSetting(Common::String *name) {
+    assert(settingMap.contains(*name));
+    setting = settingMap.getVal(*name);
+
+    debug("loading setting %s", name->c_str());
+
+    prog = (Inst *) &setting->prog;
+    stack = (Datum *) &setting->stack;
+
+    stackp = stack;
+    progp = prog;
+
+    /*for (Inst *pc_ = progp; pc_-progp < 100; pc_++) {
```

Comment:
This should be fixed, including replacing all the debug by the use of a channels. I removed that old code, since it was reading unallocated memory, but I'm not sure where exactly.